source/agora/consensus/validation/Transaction.d

```diff
@@ -132,13 +132,25 @@ public string isInvalidReason (
         auto error = checkPayload(tx);
         if (error !is null) return error;
     }
+    else if (tx.type == TxType.Coinbase)
+    {
+        if (tx.inputs.length != 1)
+            return "Transaction: Coinbase transactions must" ~
+                "include a single Input";
+
+        if (tx.inputs[0] != Input(Hash.init, height))
```

Comment:
I'm not sure this is the best idea. Couldn't this end up in a situation where there are multiple coinbase transactions (in different blocks) with the same hash? That could end up overwriting a previous utxo.

Reply:
I did this to avoid exactly that actually :) Bitcoin had that problem in early days apparently, see BIP34.

> there are multiple coinbase transactions (in different blocks) with the same hash

Wouldnt that be a hash collision anyways? Their inputs will have different heights, shouldnt they have different hashes? As long as we create single coinbase tx per address in a block, we should be fine IMO.

Reply:
I don't think the height is part of the tx hash.

Reply:
You could do a simple test. Create two exactly same coinbase txs at different heights and see if they produce unique utxos.

Reply:
Height actually goes into UTXO hash of the input.

I have tried before, but I'll add an explicit test for it anyways.

Reply:
> Height actually goes into UTXO hash of the input.

That doesn't sound right to me. Maybe we're talking about different topics?

Reply:
> I did this to avoid exactly that actually :) Bitcoin had that problem in early days apparently, see BIP34.

Yes but the way they did this is to actually include the height in the (hashing) of the coinbase transaction. But I don't see that change in your code?

Reply:
The ctor I use there gets a TX hash and output index to build an input. It builds UTXO hash with hash(TX hash, output index). I just repurposed it, the Input it creates has the utxo hash of hash(Hash.init, height). It is just a unique value for each height, but an invalid UTXO hash.

Reply:
Oh, it was a sneaky change. Could we add an explicit overload of the ctor that takes a `Height`?